### PR TITLE
Differentiate paths in a group by changing their color

### LINKED ISSUE #18: Differentiate different paths being displayed
When multiple paths are being displayed it should be clear where one ends and another starts. It would even be preferred if it was clear which was 1st, 2nd 3rd etc.
Example of 3 paths being displayed
screenshot from 2018-06-22 16-29-03

### PR DESCRIPTION
**Pull request by nickschatz**
Changes the color of paths, waypoints, and treeview items if they're in an auton group. Fixes #18

### CODE REVIEW

File: src/main/java/edu/wpi/first/pathui/FxUtils.java [hunk 1]
```diff
@@ -22,4 +25,37 @@ public static MenuItem menuItem(String text, EventHandler<ActionEvent> eventHand
     return menuItem;
   }
 
+  /**
+   * Get the color according to sub-child order.
+   * @param i The index of the child relative to the parent
+   * @return The color to draw the thing as
+   */
+  public static Color getColorForSubChild(int i) {
+    Color[] colors = new Color[]{Color.ORANGE, Color.VIOLET, Color.DARKRED};
```
Comment (AustinShalit) [inline src/main/java/edu/wpi/first/pathui/FxUtils.java]: Can you make this a static final member so we don't need to create the array every time?

File: src/main/java/edu/wpi/first/pathui/FxUtils.java [hunk 2]
```diff
@@ -22,4 +25,37 @@ public static MenuItem menuItem(String text, EventHandler<ActionEvent> eventHand
...
+  }
+
+  /**
+   * Check whether the item is a child of a child (exactly the child of a child of the root.)
+   * @param view The treeview to check against
+   * @param item The item to check
+   * @return Whether this is a sub child
+   */
+  public static boolean isSubChild(TreeView view, TreeItem item) {
+    return view.getTreeItemLevel(item) == 2;
```
Comment (AustinShalit) [inline src/main/java/edu/wpi/first/pathui/FxUtils.java]: Is there a better way to check this?

File: src/main/java/edu/wpi/first/pathui/FxUtils.java [hunk 3]
```diff
@@ -22,4 +25,37 @@ public static MenuItem menuItem(String text, EventHandler<ActionEvent> eventHand
...
+   * @param i The index of the child relative to the parent
+   * @return The color to draw the thing as
+   */
+  public static Color getColorForSubChild(int i) {
+    Color[] colors = new Color[]{Color.ORANGE, Color.VIOLET, Color.DARKRED};
+    return colors[i % colors.length];
+  }
+
+  /**
+   * Check whether the item is a child of a child (exactly the child of a child of the root.)
```
Comment (AustinShalit) [inline src/main/java/edu/wpi/first/pathui/FxUtils.java]: Period after close paren.  `root).`

File: src/main/java/edu/wpi/first/pathui/FxUtils.java [hunk 4]
```diff
@@ -22,4 +25,37 @@ public static MenuItem menuItem(String text, EventHandler<ActionEvent> eventHand
     return menuItem;
   }
 
+  /**
+   * Get the color according to sub-child order.
+   * @param i The index of the child relative to the parent
+   * @return The color to draw the thing as
+   */
+  public static Color getColorForSubChild(int i) {
+    Color[] colors = new Color[]{Color.ORANGE, Color.VIOLET, Color.FUCHSIA};
```
Comment (AustinShalit) [inline src/main/java/edu/wpi/first/pathui/FxUtils.java]: Can you make this a field of the class?

File: src/main/java/edu/wpi/first/pathui/PathDisplayController.java [hunk 5]
```diff
@@ -92,15 +93,16 @@ private void setupPathListeners() {
   }
 
   /**
-   * Add path to Controller.
-   *
-   * @param fileLocations Current working directory
-   * @param fileName      Name of path file inside directory
+   * Add a path to the controller.
```
Comment (AustinShalit) [inline src/main/java/edu/wpi/first/pathui/PathDisplayController.java]: `Adds`

File: src/main/java/edu/wpi/first/pathui/FxUtils.java [hunk 6]
```diff
@@ -22,4 +25,37 @@ public static MenuItem menuItem(String text, EventHandler<ActionEvent> eventHand
     return menuItem;
   }
 
+  /**
+   * Get the color according to sub-child order.
```
Comment (AustinShalit) [inline src/main/java/edu/wpi/first/pathui/FxUtils.java]: When writing single sentence code comments insert a hidden "The code" to the front of the sentence.  This helps make the grammar correct.

So here it would be:
The code "gets the color according to sub-child order".

TLDR: `Gets the color according to sub-child order.`

File: src/main/java/edu/wpi/first/pathui/FxUtils.java [hunk 7]
```diff
@@ -22,4 +25,37 @@ public static MenuItem menuItem(String text, EventHandler<ActionEvent> eventHand
...
+   * @param i The index of the child relative to the parent
+   * @return The color to draw the thing as
+   */
+  public static Color getColorForSubChild(int i) {
+    Color[] colors = new Color[]{Color.ORANGE, Color.VIOLET, Color.FUCHSIA};
+    return colors[i % colors.length];
+  }
+
+  /**
+   * Check whether the item is a child of a child (exactly the child of a child of the root.)
```
Comment (AustinShalit) [inline src/main/java/edu/wpi/first/pathui/FxUtils.java]: `Checks`

File: src/main/java/edu/wpi/first/pathui/FxUtils.java [hunk 8]
```diff
@@ -22,4 +25,37 @@ public static MenuItem menuItem(String text, EventHandler<ActionEvent> eventHand
...
+   * @param view The treeview to check against
+   * @param item The item to check
+   * @return Whether this is a sub child
+   */
+  public static boolean isSubChild(TreeView view, TreeItem item) {
+    return view.getTreeItemLevel(item) == 2;
+  }
+
+  /**
+   * Get the index of this item relative to its siblings.
```
Comment (AustinShalit) [inline src/main/java/edu/wpi/first/pathui/FxUtils.java]: `Gets`

File: src/main/java/edu/wpi/first/pathui/FxUtils.java (hunk 2 above)
Comment (gftabor) [reply src/main/java/edu/wpi/first/pathui/FxUtils.java]: That is probably my fault. TreeViews requires that every treeItem have the same object type which seemed cumbersome. Need to hold

Strings (The top level, text "Paths" and "Autons")
Paths (some object representing a sub child of an auton, preferably without going through the effort actually parsing the path files that are  not needed yet)
Auton (some object representing a first level child in the "Autons" treeView)

It is currently implemented where everything is just a string, either the title of the section or the filename. There has been some discussion of walking back the initial decision so easy helper functionality like
if(item instanceOf PathItem) 
could be available. Currently the only way to get what type of item is by its position in the TreeView. 
I have used
item.getParent().getParent() == null for similar checks.

Comment (nickschatz) [reply src/main/java/edu/wpi/first/pathui/FxUtils.java]: There's a lot of non-ideal coding in this that's just the result of TreeView overall not being quite built for this situation. This kind of check wouldn't be necessary if it were possible to restyle TreeItems without changing the TreeCell, for example. I'd love to have a better way, but after a good bit of searching last night this is the most elegant solution I could come up with.

**Comment by nickschatz** [comment]
image
New image, with background color.

**Comment by gftabor** [comment]
@nickschatz the other thing people requested a lot was a way of indicating what was the start and end waypoints of each path. 
Few ideas for how to do that

1. Use the color from the path before on the starting waypoint, and color from path after on the ending waypoint.
2. Have start waypoints have a Green border around the dot and a red border around the end waypoints.
3. Some sort of numbering

I would imagine I like 2 the best but am unsure.

File: src/main/java/edu/wpi/first/pathui/PathCell.java [hunk 9]
```diff
@@ -46,6 +47,32 @@ public PathCell(boolean validDropTarget, BiFunction<String, String, Boolean> val
...
+
+  private void updateColor(int index) {
+    // Determine whether this is a child of an auton (child of a child)
+    if (this.getTreeItem() != null && FxUtils.isSubChild(this.getTreeView(), this.getTreeItem())) {
+      Color color = FxUtils.getColorForSubChild(index);
+      int r = (int) (color.getRed() * 255);
+      int g = (int) (color.getGreen() * 255);
+      int b = (int) (color.getBlue() * 255);
+      String rgbStr = r + ", " + g + ", " + b;
+      this.setStyle("-fx-background-color: rgb(" + rgbStr + ");");
```
Comment (SamCarlberg) [inline src/main/java/edu/wpi/first/pathui/PathCell.java]: I wouldn't mix CSS and code like this. You could either set a CSS pseudoclass eg `path-cell:first`, `path-cell:second`, or call `setBackground(new Background(new BackgroundFill(color, null, null))`

File: src/main/java/edu/wpi/first/pathui/FxUtils.java [hunk 10]
```diff
@@ -8,7 +8,15 @@
 import javafx.scene.paint.Color;
 
 public final class FxUtils { // NOPMD util class name
-  private static Color[] childColors = new Color[] {Color.ORANGE, Color.VIOLET, Color.FUCHSIA};
```
Comment (AustinShalit) [inline src/main/java/edu/wpi/first/pathui/FxUtils.java]: final please

**Comment by gftabor** [comment]
screenshot from 2018-07-25 13-48-00
After the most recent changes to use 8 colors from JavaFX charts.

**Comment by nickschatz** [comment]
Current looks, with red and green outlines
image

**Comment by SamCarlberg** [comment]
I think the solid border on the end points is too much. A colored glow or dropshadow effect would be better

**Comment by nickschatz** [comment]
image
With triangles.

File: src/main/java/edu/wpi/first/pathui/Waypoint.java [hunk 11]
```diff
@@ -28,7 +28,9 @@
 
 
   private final Line tangentLine;
-  private final Circle dot;
+  private Polygon dot;
```
Comment (kjrobrien) [inline src/main/java/edu/wpi/first/pathui/Waypoint.java]: This should no longer be called dot. Maybe "icon"?

File: src/main/resources/edu/wpi/first/pathui/style.css [hunk 12]
```diff
@@ -24,3 +22,27 @@ Line {
...
+}
+:subchild5 {
+    -fx-background-color: #9a42c8;
+}
+:subchild6 {
+    -fx-background-color: #c84164;
+}
+:subchild7 {
+    -fx-background-color: #888888;
+}
```
Comment (kjrobrien) [inline src/main/resources/edu/wpi/first/pathui/style.css]: Is this a repeat of CHILD_COLORS in FxUtils.java?

File: src/main/java/edu/wpi/first/pathui/FxUtils.java [hunk 13]
```diff
@@ -22,4 +33,57 @@ public static MenuItem menuItem(String text, EventHandler<ActionEvent> eventHand
...
+    return SUBCHILD_CLASSES[i % SUBCHILD_CLASSES.length];
+  }
+
+  /**
+   * Applies SUBCHILD_CLASSES to the specified node.
+   * @param node The node to apply the classes to
+   */
+  public static void applySubchildClasses(Node node) {
+    for (PseudoClass pc : FxUtils.SUBCHILD_CLASSES) {
+      node.getStyleClass().add(pc.getPseudoClassName());
```
Comment (SamCarlberg) [inline src/main/java/edu/wpi/first/pathui/FxUtils.java]: No. A pseudoclass is a selector, not a class in and of itself.

File: src/main/resources/edu/wpi/first/pathui/style.css [hunk 14]
```diff
@@ -2,19 +2,50 @@
     -fx-background-color: #101824;
 }
 
-Circle {
-    -fx-fill: orange;
+:subchild0 {
```
Comment (SamCarlberg) [inline src/main/resources/edu/wpi/first/pathui/style.css]: This should be

```css
.tree-cell:subchild0 {
    -fx-background-color: CHART_COLOR_2;
}

CubicCurve:subchild0 {
    -fx-stroke: CHART_COLOR_2;
}
```

I recommend setting a custom style class for the splines (maybe `path`?) instead of using the class selector `CubicCurve`, as well as creating your own variables for the colors for better naming.

**Comment by gftabor** [comment]
The colored Cells have strange behavior if you expand and unexpand that autonomous.
image
image

File: src/main/java/edu/wpi/first/pathweaver/FxUtils.java [hunk 15]
```diff
@@ -9,12 +11,14 @@
...
+  private static final List<PseudoClass> SUBCHILD_SELECTORS; // NOPMD
 
   static {
-    for (int i = 0; i < SUBCHILD_SELECTORS.length; i++) {
-      SUBCHILD_SELECTORS[i] = PseudoClass.getPseudoClass("subchild" + i);
+    PseudoClass[] selectors = new PseudoClass[8];
+    for (int i = 0; i < selectors.length; i++) {
+      selectors[i] = PseudoClass.getPseudoClass("subchild" + i);
     }
+    SUBCHILD_SELECTORS = List.of(selectors);
```
Comment (AustinShalit) [inline src/main/java/edu/wpi/first/pathweaver/FxUtils.java]: Use `Collections.unmodifiableList`

Comment (nickschatz) [reply src/main/java/edu/wpi/first/pathweaver/FxUtils.java]: Why? I originally did use that and IntelliJ suggested replacing it with `List.of`. It seems that both are immutable, and using `List.of` doesn't require calling `Arrays.asList` first.

Comment (SamCarlberg) [reply src/main/java/edu/wpi/first/pathweaver/FxUtils.java]: `List.of` is the preferred method starting in Java 9

Comment (AustinShalit) [reply src/main/java/edu/wpi/first/pathweaver/FxUtils.java]: Oh.  You are right
